### ISSUE TITLE
Infer parameter types from base class methods

### ISSUE THREAD

**Issue by relsunkaev** [microsoft/pyright#2982]
When inheriting from an `ABC` with abstract methods, defining these abstract methods still requires specifying their parameter types. If one or several of these parameter types needs to be imported or, even worse, only imported when `TYPE_CHECKING == TRUE` (which I would imagine is not uncommon) this can get kind of clunky where you have to insert that import atop every file that uses the `ABC`.

Since the parameter specification (and other types) for an implementation of an abstract method have to conform to the one defined in the `ABC`, am I correct to assume that these parameter types could potentially be inferred from the `ABC` or are there concerns associated with doing that?

**Comment by erictraut** [comment]
It's an interesting idea. 

I can think of a number of conditions that would need to be considered.

1. The parameter signature of the child class doesn't match the parameter signature of the parent class (arity mismatch, *args or **kwargs mismatch, parameter name mismatch, etc.). Perhaps in this case, inference is not used — i.e. it defaults back to the current behavior where parameter types are Unknown.

2. A subset of the child parameters are annotated (either matching the annotations of the parent method or presenting a type mismatch). It probably makes sense in this case for inference not to be used for any parameters.

3. The method in the parent class is declared with overloads. If there is an annotated implementation in the parent class, that signature could be used as a reference. If there is no implementation (e.g. the base class is defined in a type stub and has no implementation), then inference would probably not be possible.

4. The base class is generic and some of the parameters contain class-scoped TypeVars. In this case, I guess the base class TypeVars could be specialized according to the child class declaration.

5. The base class method is generic and has function-scoped TypeVars. In this case, I suppose a new function-scoped TypeVar could be assumed.

Is there a reason you think this would apply only to ABCs? Why wouldn't is apply any time a child class overrides a method in a base class?

**Comment by erictraut** [comment]
I thought of a few additional conditions and considerations:

6. The method in the base class has a subset of its parameters annotated but one or more are missing. In this case, only the types of the corresponding parameters should be inferred.

7. The method in the base class has one or more default arguments but the override does not. In this case, it should probably be considered a signature mismatch and parameter type inference shouldn't be used.

8. The method in the child class has default arguments specified for one or more parameters and the type of the default argument expression isn't type compatible with the annotated type of the corresponding parameter in the base class. I'm not sure what to do about this condition. Should the parameter types for the method still be inferred in this case? If so, should a type error be reported? This one is problematic.

9. If the user has the `reportMissingParameterType` diagnostic check enabled, should it emit a diagnostic if a parameter type is not annotated but its type can be inferred from the base class signature? Since `reportMissingParameterType` is a strict-mode option, I think it should emit a diagnostic in this case.

**Comment by hmc-cs-mdrissi** [comment]
I'm -0 on this especially for subclassing classes from external libraries. Many libraries I use the original class if type annotated well has very broad input signature. Having a broad signature is great as a user of the library. Libraries like numpy converts so many things for you. But most of the time when I subclass them I only include a subset of the types that my code plans to handle. Toy example,

```python
class LibFoo
  def bar(x: Iterable[float | str]):
    ...

class UserFoo(LibFoo):
  def bar(x: List[float]):
    ...
```

is pretty common. The first type is better, but the second type may be easier to implement and broadness of first type unneeded for my usage.

For classes within your own codebase I think this would be reasonably accurate.

**Comment by relsunkaev** [comment]
# Comment
1. Agreed mostly except for a few points:
    - Positional-only arguments could be allowed to have arbitrary names.
    - Keyword-only arguments could be allowed to be declared in arbitrary order.
8. I lean toward it inferring the parameter type from the default value provided in the child method and reporting an error if that inferred type does not conform to the one defined in the base class. The default argument in the overriding method has a closer proximity to the code being looked at (doesn't get any closer than right next to it) and is far more likely to be the type "intended" by whoever is doing the overriding.
9. Agreed. If it emits a diagnostic when it can infer types from an `overload` often placed right above it, it should definitely emit one when inferring them from a base class that is at best located in a neighboring module. All for being explicit when on strict.

All others I agree with completely.

As for the focus on ABCs, there is no reason. I just happened to be  working with an ABC when I encountered this. Don't know why I got so hung up on it being an ABC.

**Comment by relsunkaev** [comment]
> But most of the time when I subclass them I only include a subset of the types that my code plans to handle.

Doing this would imply your child class is incompatible with the protocol defined by its parent.

If I write a method that takes in a `LibFoo`, the protocol tells me I should be able to pass any `Iterable` to its `bar` method and so I pass a `tuple`. Now, ignoring the `float | str`->`float` constraint, `UserFoo.bar()` takes in a `list` which tells me it uses `list`-specific methods like `append`. This means that passing a `UserFoo` to my method, which I allege it support, would result in a runtime error. `UserFoo` "lied" about being a `LibFoo` because you can't call its `bar` method with a`tuple`. These scenarios are hard to debug in complex codebases and are why the [Liskov substitution principle](https://en.m.wikipedia.org/wiki/Liskov_substitution_principle) was proposed.

Of course, if `UserFoo.bar()` doesn't use any `list`-specific methods the issue goes away, but specifying the type as `list` becomes unnecessarily restrictive.

**Comment by hmc-cs-mdrissi** [comment]
Yes I agree. But usually user code only works on a much smaller scope. tensorflow as an example lets you mix numpy arrays, lists, tuples, tensors, and scalars in a variety of ways. It's quite rare for user code written in tensorflow to try to have all those types handled. Or numpy arrays allowing string/datetime arrays, but most arrays people work with assumes float/int (sometimes only float).

I think Liskov failures are closer norm when subclasses 3rd party libraries and not uncommon case. Maybe this is my bias working with mostly numerical libraries that have extremely broad input types.

**Comment by erictraut** [comment]
@hmc-cs-mdrissi, as you said, this feature wouldn't be interesting to users like you who want strict type checking. Given the rules I've outlined above, it wouldn't affect you because you enable `reportMissingParameterType`. It also wouldn't affect my team's code base because we enable all of the strictest type checking settings including `reportMissingParameterType`. But I can understand why this feature might be appealing to some pyright users.

I'm considering adding this primarily for pylance users, most of whom use no type annotations in their code but still want accurate completion suggestions. If pyright adds support for parameter type inference for method overrides, it would improve the experience for this large group of users. It would also likely improve inference of return types for functions and methods in unannotated libraries.

**Comment by relsunkaev** [comment]
It would also be very helpful for people moving to strict mode (like myself). Instead of being inundated with "type is Unknown" errors it will just highlight the relevant part and make it clear that all that needs to be done is adding type annotations for the parameters. Switching into strict and seeing the problem count shoot up into 4 digits because of things like this can be very daunting.

**Comment by erictraut** [comment]
I've implemented a first cut at this functionality. It will be in the next release of pyright and pylance. Here's the documentation: https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#parameter-type-inference.

**Comment by erictraut** [comment]
This is included in pyright 1.1.224, which I just published. It will also be included in the next release of pylance.
